modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/util/RexUtils.java

```diff
@@ -176,6 +182,104 @@ public static boolean isIdentity(List<? extends RexNode> projects, RelDataType i
+            List<RexNode> operands;
+            switch (rex.getKind()) {
+                case AND:
+                    operands = flattenAnd(((RexCall) rex).getOperands());
+                    RexNode head = operands.get(0);
+                    RexNode headDnf = toDnf(head);
+                    List<RexNode> headDnfs = RelOptUtil.disjunctions(headDnf);
+                    RexNode tail = and(Util.skip(operands));
+                    RexNode tailDnf = toDnf(tail);
+                    List<RexNode> tailDnfs = RelOptUtil.disjunctions(tailDnf);
+                    List<RexNode> list = new ArrayList<>(headDnfs.size() * tailDnfs.size());
+                    for (RexNode h : headDnfs) {
+                        for (RexNode t : tailDnfs) {
+                            list.add(and(ImmutableList.of(h, t)));
+                        }
+                    }
+                    return or(list);
+                case OR:
+                    operands = flattenOr(((RexCall) rex).getOperands());
+                    List<RexNode> processed = toDnfs(operands);
+                    return or(processed);
+                case NOT:
+                    RexNode arg = ((RexCall) rex).getOperands().get(0);
+                    switch (arg.getKind()) {
+                        case NOT:
+                            return toDnf(((RexCall) arg).getOperands().get(0));
+                        case OR:
+                            operands = ((RexCall) arg).getOperands();
+                            return toDnf(
+                                    and(Util.transform(flattenOr(operands), RexUtil::not)));
+                        case AND:
+                            operands = ((RexCall) arg).getOperands();
+                            return toDnf(
+                                    or(Util.transform(flattenAnd(operands), RexUtil::not)));
+                        default:
+                            return rex;
+                    }
+                default:
+                    return rex;
+            }
+        }
+
+        private List<RexNode> toDnfs(List<RexNode> nodes) {
+            List<RexNode> list = new ArrayList<>();
+            for (RexNode node : nodes) {
+                RexNode dnf = toDnf(node);
+                switch (dnf.getKind()) {
+                    case OR:
+                        list.addAll(((RexCall) dnf).getOperands());
+                        break;
+                    default:
+                        list.add(dnf);
+                }
+            }
+            return list;
+        }
+
+        private RexNode and(Iterable<? extends RexNode> nodes) {
+            return composeConjunction(rexBuilder, nodes);
+        }
+
+        private RexNode or(Iterable<? extends RexNode> nodes) {
+            RexNode res = composeDisjunction(rexBuilder, nodes);
+            if (res instanceof RexCall) {
+                RexCall res0 = (RexCall) res;
+                if (res0.getOperands().size() > maxOrNodes) {
```

Comment:
Is this condition the only difference between this code and RexUtils::DnfHelper?

Reply:
yes, on planning phase with huge amount of params it takes more time than planning is available, thus planning timeout will raised